lib/core/src/chain_swap.rs

```diff
@@ -681,148 +723,298 @@ impl ChainSwapStateHandler {
+                err: format!(
+                    "Unexpected refund tx type returned for outgoing Chain swap {}",
+                    swap.id
+                ),
+            });
+        };
+        let refund_tx_id = liquid_chain_service
+            .broadcast(&refund_tx, Some(&swap.id))
+            .await?
+            .to_string();
+
+        info!(
+            "Successfully broadcast refund for outgoing Chain Swap {}, is_cooperative: {is_cooperative}",
+            swap.id
+        );
+
+        Ok(refund_tx_id)
+    }
+
+    async fn check_swap_expiry(&self, swap: &ChainSwap) -> Result<bool> {
+        let swap_creation_time = UNIX_EPOCH + Duration::from_secs(swap.created_at as u64);
+        let duration_since_creation_time = SystemTime::now().duration_since(swap_creation_time)?;
+        if duration_since_creation_time.as_secs() < 60 * 10 {
+            return Ok(false);
+        }
+
+        match swap.direction {
+            Direction::Incoming => {
+                let swap_script = swap.get_lockup_swap_script()?.as_bitcoin_script()?;
+                let current_height = self.bitcoin_chain_service.lock().await.tip()?.height as u32;
+                let locktime_from_height =
+                    LockTime::from_height(current_height).map_err(|e| PaymentError::Generic {
+                        err: format!("Error getting locktime from height {current_height:?}: {e}",),
+                    })?;
+
+                info!("Checking Chain Swap {} expiration: locktime_from_height = {locktime_from_height:?},  swap_script.locktime = {:?}", swap.id, swap_script.locktime);
+                Ok(swap_script.locktime.is_implied_by(locktime_from_height))
+            }
+            Direction::Outgoing => {
+                let swap_script = swap.get_lockup_swap_script()?.as_liquid_script()?;
+                let current_height = self.liquid_chain_service.lock().await.tip().await?;
+                let locktime_from_height = ElementsLockTime::from_height(current_height)?;
+
+                info!("Checking Chain Swap {} expiration: locktime_from_height = {locktime_from_height:?},  swap_script.locktime = {:?}", swap.id, swap_script.locktime);
+                Ok(utils::is_locktime_expired(
+                    locktime_from_height,
+                    swap_script.locktime,
+                ))
+            }
+        }
+    }
+
+    pub(crate) async fn track_refunds_and_refundables(&self) -> Result<(), PaymentError> {
+        let pending_swaps = self.persister.list_pending_chain_swaps()?;
+        for swap in pending_swaps {
+            if swap.refund_tx_id.is_some() {
+                continue;
+            }
+
+            match swap.direction {
+                // Track refunds
+                Direction::Outgoing => {
+                    let refund_tx_id_result: Result<String, PaymentError> = match swap.state {
+                        Pending => self.refund_outgoing_swap(&swap, false).await,
```

Comment:
Seems like we refund any pending outgoing? without checking expiry?

Reply:
The expiry check happens within `refund_outgoing_swap`. The flow is as follows:

- Pending state: We refund non-cooperatively, which will probably fail the expiry check and return (unless it's been there past the expiry, then it will refund)

- Refund Pending: We first refund cooperatively, then non-cooperatively. In the second case we check for expiry.

Reply:
While the check inside refund_outgoing_swap is in its place IMO, I think it is a bit misleading to call refund_outgoing_swap for every pending (for RefundPending it seems OK).
I think we should check for expiry before calling the non cooperative refund on pending (or fetch only the expired ones from the db).

Reply:
Also note that the refund_outgoing_swap throws error in case it is not expired so we will get unnecessary errors in the logs.